config/application.rb

```diff
@@ -18,7 +18,7 @@
 module Collections
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
-    config.load_defaults 6.0
+    config.load_defaults 6.1
```

Comment:
Any concerns about the backwards compatibility of the new defaults?

Reply:
I've tested with the following configuration (i.e. pointing to 6.0, and then this code explicitly in config/initializers):

```ruby
# Be sure to restart your server when you modify this file.
#
# This file contains migration options to ease your Rails 6.1 upgrade.
#
# Once upgraded flip defaults one by one to migrate to the new default.
#
# Read the Guide for Upgrading Ruby on Rails for more info on each option.

# Support for inversing belongs_to -> has_many Active Record associations.
# Rails.application.config.active_record.has_many_inversing = true

# Track Active Storage variants in the database.
# Rails.application.config.active_storage.track_variants = true

# Apply random variation to the delay when retrying failed jobs.
Rails.application.config.active_job.retry_jitter = 0.15

# Stop executing `after_enqueue`/`after_perform` callbacks if
# `before_enqueue`/`before_perform` respectively halts with `throw :abort`.
Rails.application.config.active_job.skip_after_callbacks_if_terminated = true

# Specify cookies SameSite protection level: either :none, :lax, or :strict.
#
# This change is not backwards compatible with earlier Rails versions.
# It's best enabled when your entire app is migrated and stable on 6.1.
Rails.application.config.action_dispatch.cookies_same_site_protection = :lax

# Generate CSRF tokens that are encoded in URL-safe Base64.
#
# This change is not backwards compatible with earlier Rails versions.
# It's best enabled when your entire app is migrated and stable on 6.1.
Rails.application.config.action_controller.urlsafe_csrf_tokens = true

# Specify whether `ActiveSupport::TimeZone.utc_to_local` returns a time with an
# UTC offset or a UTC time.
ActiveSupport.utc_to_local_returns_utc_offset_times = true

# Change the default HTTP status code to `308` when redirecting non-GET/HEAD
# requests to HTTPS in `ActionDispatch::SSL` middleware.
Rails.application.config.action_dispatch.ssl_default_redirect_status = 308

# Use new connection handling API. For most applications this won't have any
# effect. For applications using multiple databases, this new API provides
# support for granular connection swapping.
# Rails.application.config.active_record.legacy_connection_handling = false

# Make `form_with` generate non-remote forms by default.
Rails.application.config.action_view.form_with_generates_remote_forms = false

# Set the default queue name for the analysis job to the queue adapter default.
# Rails.application.config.active_storage.queues.analysis = nil

# Set the default queue name for the purge job to the queue adapter default.
# Rails.application.config.active_storage.queues.purge = nil

# Set the default queue name for the incineration job to the queue adapter default.
# Rails.application.config.action_mailbox.queues.incineration = nil

# Set the default queue name for the routing job to the queue adapter default.
# Rails.application.config.action_mailbox.queues.routing = nil

# Set the default queue name for the mail deliver job to the queue adapter default.
# Rails.application.config.action_mailer.deliver_later_queue_name = nil

# Generate a `Link` header that gives a hint to modern browsers about
# preloading assets when using `javascript_include_tag` and `stylesheet_link_tag`.
Rails.application.config.action_view.preload_links_header = true
```

Despite the new CSRF and preload_links_header settings, I don't see any difference in the DOM output comparing http://localhost:3070/coronavirus and https://www.gov.uk/coronavirus.

Reply:
Great, thanks for checking - I don't think this app has any post routes so CSRF may not impact it at all.

I understand we had some problems with the preload link header on some GOV.UK apps as it can suggest preloading assets that are behind conditional IE comments - I understand GOV.UK may well have been the [motivation for it to become an option](https://github.com/rails/rails/pull/40882). I'm assuming it's also not part of the DOM output but instead a HTTP header in the response.

I just gave it a spin on integration to try learn but it bombed example error: https://sentry.io/organizations/govuk/issues/2276420286/?environment=integration-blue-aws&project=202213&query=is%3Aunresolved - which I think is also why it's failing e2e tests.

Reply:
Looks as though there's an upstream bug in dalli: https://github.com/petergoldstein/dalli/issues/771

...which affects Collections, Government Frontend, Whitehall and Finder Frontend: https://github.com/search?q=org%3Aalphagov+config.cache_store+%3D+%3Adalli_store&type=code

Looks like the fix is relatively trivial: swap out Dalli store for MemCacheStore: https://github.com/alphagov/local-links-manager/pull/751

I'll push up a commit and build this to integration to test.

Reply:
That's fixed the E2E, @kevindew 😄

Reply:
You're right that the preload change comes in a header. Visited https://www.gov.uk/coronavirus on Integration and had this `link` header response:

```
link: </assets/collections/application-158e5fb957e57cf631a8e97f298a189c29433a89f096c83adb8a3146a410b944.js>; rel=preload; as=script; nopush,</assets/collections/application-d298fc8cb919179a40beac19e17c539d2e9d0323422281ca3c36f41933d05fa5.css>; rel=preload; as=style; nopush,</assets/collections/print-f23cb8ccbf886494b66cbdb4055804d044ddbec2d80f7f05f3f6a4c18aef696b.css>; rel=preload; as=style; nopush
```

It doesn't seem to pull in IE-conditional resources, thankfully.

Reply:
Oh great - I think this app doesn't have any IE conditional resources (those are done by static).

Awesome work on the dalli issue